src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -317,9 +317,12 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
           }
           else {
             mnode.visitVarInsn(asm.Opcodes.ALOAD, 0)
-            generatedType =
-              if (tree.symbol == ArrayClass) ObjectRef
```

Comment:
This special case is old, it goes back to https://github.com/scala/scala/commit/0ad5e129f3. Note that on a current 2.12.x, that code path is never triggered when compiling library+compiler+reflect. With this PR we get here again, but `Object` doesn't work when the expected type is `[Object`.